### ISSUE TITLE
Change no-use-before-define default behaviour for functions

### ISSUE THREAD

**Issue by hamboomger** [eslint/eslint#13261]
**What rule do you want to change?**
no-use-before-define
**Does this change cause the rule to produce more or fewer warnings?**
This change will cause fewer amount of warnings
**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behaviour - disable the `no-use-before-define` for functions by default, because it seems like there is no reason to have it enabled. Quote from the `no-use-before-define` rule page itself:

> functions (boolean) - The flag which shows whether or not this rule checks function declarations. ... . **Function declarations are hoisted, so it's safe**. Default is true.

Why is it true then? Having private functions above public makes code less readable, and defining private functions below public ones is the idiomatic way of writing code in many languages, and  [it's not only my opinion](https://stackoverflow.com/questions/1760850/best-practice-ordering-of-public-protected-private-within-the-class-definition), ([aand another link](https://stackoverflow.com/a/310967/5362302)).
**Please provide some example code that this change will affect:**

```js
// named export functions declared on top to make it easy
// to discover the public API exposed by the module
export function namedPublicMethod() {
  // do things
  return privateHelperFunction();
}

// less often read private utility functions declared on the bottom of the file
function privateHelperFunction() {
  // do things
}
```
Code was used from the similar issue [here](https://github.com/FormidableLabs/eslint-config-formidable/issues/31#issue-182083012)
**What does the rule currently do for this code?**
Currently this code will produce an error, because the `privateHelperFunction` was used before it was declared.
**What will the rule do after it's changed?**
It will not produce the warning anymore.
**Are you willing to submit a pull request to implement this change?**
Yep

**Comment by anikethsaha** [comment]
Thanks for the issue. 

**IMO**, In this above example, currently there is no way to tell which function is private or which one is not. also, access specifier is available only for classes (as methods )  and also the [proposal is still in Stage 3](https://github.com/tc39/proposal-private-methods) and eslint doesn't support stage-3 concepts in core rules. 

>  Having private functions above public makes code less readable, and defining private functions below public ones is the idiomatic way of writing code in many languages

Umm, I don't think I can agree in this, In most OOP languages (C++ for instances ), I have personally and saw others use private first. 

_(though its irrelevant cause javascript doesnt support private as of now , still )_

The purpose of this rule is to avoid usage before define.

> This can be confusing and some believe it is best to always declare variables and functions before using them.

so , I am personally not in favor of this change. 

lets hear for others!

**Comment by hamboomger** [comment]
> **IMO**, In this above example, currently there is no way to tell which function is private or which one is not. 
I think any exported function can be considered public and every other function can be considered private in scope of a file, I was not talking about ES6 classes, but rather about functions in scope of a file (bad terminology from my side, maybe)
> The purpose of this rule is to avoid usage before define.
But why should we avoid functions declarations after their usage? Is it considered a bad practice? It's safe to use, because function declarations are hoisted.

**Comment by anikethsaha** [comment]
> But why should we avoid functions declarations after their usage? Is it considered bad practice? It's safe to use, because function declarations are hoisted.

it's good for code readability.  also, the hoisting is not true for anonymous  (`FunctionExpression`) functions. 

they hoist the name instead of assignment (def).


[Refer this](https://github.com/airbnb/javascript#hoisting--anon-expressions)

**Comment by hamboomger** [comment]
@anikethsaha anonymous functions are considered as variables by this ESLint rule, and I'm proposing to disable the rule for functions only - I fully agree that you should not use anonymous functions before declaration. So even if you disable the `no-use-before-define` rule for functions as I propose, ESLint will still throw an error if you try to use anonymous functions in such way

**Comment by anikethsaha** [comment]
I would suggest to use `/*eslint no-use-before-define: ["error", { "functions": false }]*/` for this.

lets see what others think 👍

**Comment by mdjermanovic** [comment]
It isn't always safe. For example, this would throw on `f()`:

```js
f();

const a = 1;

function f() {
  return a;
}
```

I'd personally expect from this rule to enforce "no use before define" on everything by default, which is the actual behavior.

**Comment by nzakas** [comment]
Yeah, I don't see us changing the default behavior on this rule, which is one of our oldest. There are plenty of options for changing how the rule behaves, including using a disable comment. If you want to create your own version, I'd suggest looking at [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer), which lets you run an existing rule and then filter out the messages you don't want to see.

Thanks for the suggestion, but closing as we won't be making this change.
